### PR TITLE
Add a "copy URL" button in the preview toolbar

### PR DESCRIPTION
**Pull request by simonreitinger**
This PR adds the real URL of the currently displayed page with a "copy to clipboard" button to the preview toolbar. 
I did not come up with a solution how to handle query parameters like page or news. Any ideas maybe?

Bildschirmfoto 2020-02-10 um 14 01 18

### CODE REVIEW

**Comment by Toflar** [comment]
Thanks for contributing! Can you explain why you need that? You can just copy the address from the toolbar? Or do you want to be able to copy the non-preview URL?
And why do we need to show the URL? A button would be enough, no? :)

**Comment by simonreitinger** [comment]
Yeah, this is for the real URL. Copying it from the toolbar is actually just a shortcut instead of publishing the page in the back end and so on. I thought this would be suitable as a core feature. Otherwise I can put this in a seperate Contao extension. Button only should be fine, too!

**Comment by Toflar** [comment]
I don't understand why one would only want to copy the path of the request and not the full URI?

**Comment by simonreitinger** [comment]
It actually is the full URI:
https://github.com/contao/contao/pull/1301/files#diff-3ce6b762ee02232a8b63419909fd0815R110

Maybe I should rename the query parameter 🤔

**Comment by Toflar** [comment]
No `getPathInfo()` only returns - well, the path info :) But you also want the whole query parameters (`$request->getUri()`). And the request is already passed on to the template so I don't understand why you need another parameter. Well, if you want to get the real URI (so without `preview.php` you maybe need another parameter because it's likely easier to strip it in PHP than in the twig template.

**Comment by Xirdion** [comment]
Instead of adding the real path as GET parameter you could use it in the JS-Code and set the value of the new input field.
This way you don't need the request in the BackendPreviewSwitchController.
image

**Comment by frontendschlampe** [comment]
Just a question: Why you need the real URL?

If you make some changes which will be hidden, you need the preview URL otherwise nobody see the changes. If you make changes which won't be hidden, you don't need the preview mode.

So I think, this feature isn't necessary at all, isn't it?

**Comment by Xirdion** [comment]
It would be a feature for our customers. 
We are telling them that they should check new articles in the preview mode to see how the result will look like.
But then they are using the URL in the browser with the `preview.php` to prepare a new facebook post or something like this. We are telling them that they have to remove the preview.php but .. well, it seems like this is not that easy ;)
So the reason for this feature is to point our customers an easy way out to get the real URL of the page.

**Comment by Toflar** [comment]
I'm not against such a button. But it can be a simple button that just copies the public URL.
This PR is way too complex. You already have the complete `Request` instance in the template. All you need to do is to also pass the preview script (`preview.php` is configurable) to the template and have JavaScript replace it from `window.location.href`?

**Comment by simonreitinger** [comment]
Let's try again 😄 

I guess a visible input is needed as `document.execCommand('copy')` does not support hidden inputs. 

We could use the [Clipboard API](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API) for that but Edge, Safari and Opera do not support it (yet).

**Comment by Toflar** [comment]
Afaik it's discouraged to use `document.execCommand()` but yeah, we maybe need a polyfill to properly deal with the Clipboard API in all browsers.
But wasn't the initial idea to have a button that copies the real URL (so the whole URI including protocol, host etc. just without the `preview.php` chunk)?

**Comment by Xirdion** [comment]
But `window.location.href` contains the whole URL with protocol and host. We are just removing the part with **/preview.php** with `window.location.href.replace('{{ app.request.baseUrl }}', '');`.

**Comment by Toflar** [comment]
Yeah but that's the point. The script is configurable. I can also configure it to `foobar/super-preview.php`.

**Comment by simonreitinger** [comment]
I replaced the base path with the preview script string which already is injected in `PreviewToolbarListener`. This works with other entrypoints than "preview.php" as well.

**Review by Toflar — COMMENTED** [review]
You can fix the coding style issues running `composer cs-fixer -- --fix`.
@contao/developers LGTM, wdyt? @ausi wdyt regarding the use of `document.execCommand()`? Seems like it's still the most widely supported option and I don't feel like loading a polyfill in the preview toolbar.

**Comment by ausi** [comment]
I would prefer to use the new clipboard API if available and fallback to the `execCommand` approach otherwise:

```js
copyButton.addEventListener('click', function(event) {
    function clipboardFallback() {
        input.select();
        input.setSelectionRange(0, 99999);
        document.execCommand('copy');
    }
    event.preventDefault();
    if (navigator.clipboard && navigator.clipboard.writeText) {
        navigator.clipboard.writeText(input.value).catch(clipboardFallback);
    }
    else {
        clipboardFallback();
    }
});
```

**Review by ausi — DISMISSED** [review]

**Review by Toflar — DISMISSED** [review]

File: core-bundle/src/Resources/views/Frontend/preview_toolbar_base_js.html.twig [hunk 1]
```diff
@@ -284,5 +284,26 @@
...
+                    input.select();
+                    input.setSelectionRange(0, 99999);
+                    document.execCommand('copy');
+                }
+                event.preventDefault();
+                if (navigator.clipboard && navigator.clipboard.writeText) {
+                    navigator.clipboard.writeText(input.value).catch(clipboardFallback);
+                }
+                else {
+                    clipboardFallback();
```
Comment (leofeyer) [inline core-bundle/src/Resources/views/Frontend/preview_toolbar_base_js.html.twig:318]: Why is this encapsulated in a function instead of just executed right here?

Comment (ausi) [reply core-bundle/src/Resources/views/Frontend/preview_toolbar_base_js.html.twig:318]: Because it is used twice, see end of line 301.

Comment (leofeyer) [reply core-bundle/src/Resources/views/Frontend/preview_toolbar_base_js.html.twig:318]: My bad. 🙈

**Review by leofeyer — CHANGES_REQUESTED** [review]
I like the general idea, however, I think we have to improve the UX. This is how it currently looks:





Here is what bothers me:

1. The input field is so small that I can only see the domain.
2. The button only says "Copy", however, it is not clear what I am copying.

I would suggest that we remove the input field altogether and rename the button to "Copy URL".

**Comment by aschempp** [comment]
didn't @richardhj want to give an overhaul to the preview toolbar anyway? 🙃

**Comment by aschempp** [comment]
how do we continue here?

**Comment by Aybee** [comment]
I also very often remove the fragment "preview.php/" from the Browsers URL bar. So a button or link to open the real URL directly in the browser within the tab or with CTRL within a new tab would be very convenient. But I don't need to copy the real URL.

**Comment by leofeyer** [comment]
> how do we continue here?



@simonreitinger Are you still working on this PR?

**Comment by simonreitinger** [comment]
@leofeyer I guess a button-only solution would be the cleanest approach here. I can update the PR in the next few days 👍

File: core-bundle/src/Resources/views/Frontend/preview_toolbar_base_js.html.twig [hunk 2]
```diff
@@ -299,5 +299,26 @@
                 });
             });
         });
+
+        // copy published link to clipboard
+        window.addEventListener('cto-toolbar-loaded', function() {
+            const copyButton = document.getElementById('copyPublishedPath');
+            copyButton.addEventListener('click', function(event) {
+                const input = document.createElement('input');
+                input.value = window.location.href.replace('{{ preview_script }}', '');
```
Comment (Toflar) [inline core-bundle/src/Resources/views/Frontend/preview_toolbar_base_js.html.twig]: This won't work for URLs like `https://www.domain.com/preview.php/page.html?filter=stuff/containing/preview.php/etc.`. I know it's super unlikely but it could happen and we must prevent this from failing :)
We should only replace the first occurence right after the host.

Comment (simonreitinger) [reply core-bundle/src/Resources/views/Frontend/preview_toolbar_base_js.html.twig]: @Toflar actually this should work as expected! `String.prototype.replace()` only replaces the first occurence. For replacing every occurence there is the newer `String.prototype.replaceAll()` that was introduced some months ago.

Bildschirmfoto 2020-10-19 um 12 31 15

Comment (Toflar) [reply core-bundle/src/Resources/views/Frontend/preview_toolbar_base_js.html.twig]: Interesting. Good then :)

File: core-bundle/src/Resources/views/Frontend/preview_toolbar_base_js.html.twig [hunk 3]
```diff
@@ -299,5 +299,26 @@
...
+                function clipboardFallback() {
+                    input.select();
+                    input.setSelectionRange(0, 99999);
+                    document.execCommand('copy');
+                }
+                event.preventDefault();
+                if (navigator.clipboard && navigator.clipboard.writeText) {
+                    navigator.clipboard.writeText(input.value).catch(clipboardFallback);
```
Comment (ausi) [inline core-bundle/src/Resources/views/Frontend/preview_toolbar_base_js.html.twig]:
```suggestion
                const href = window.location.href.replace('{{ preview_script }}', '');
                function clipboardFallback() {
                    const input = document.createElement('input');
                    input.value = href;
                    document.body.appendChild(input);
                    input.select();
                    input.setSelectionRange(0, 99999);
                    document.execCommand('copy');
                    document.body.removeChild(input);
                }
                event.preventDefault();
                if (navigator.clipboard && navigator.clipboard.writeText) {
                    navigator.clipboard.writeText(href).catch(clipboardFallback);
```

I think the input has to be injected into the document in order for the `copy` command to work.

Comment (simonreitinger) [reply core-bundle/src/Resources/views/Frontend/preview_toolbar_base_js.html.twig]: @ausi true! I just implemented that. Also tested with Safari and "async clipboard API" disabled.

File: core-bundle/src/Resources/contao/languages/de/default.xlf [hunk 4]
```diff
@@ -2681,6 +2681,10 @@
         <source>Close the Contao toolbar</source>
         <target>Contao-Toolbar schließen</target>
       </trans-unit>
+      <trans-unit id="MSC.copyURL">
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/languages/de/default.xlf]: please remove the german translation from the PR, these are managed on transifex.com

File: core-bundle/src/Resources/views/Frontend/preview_toolbar_base_js.html.twig [hunk 5]
```diff
@@ -299,5 +299,29 @@
                 });
             });
         });
+
+        // copy published link to clipboard
+        window.addEventListener('cto-toolbar-loaded', function() {
+            const copyButton = document.getElementById('copyPublishedPath');
+            copyButton.addEventListener('click', function(event) {
+                const href = window.location.href.replace('{{ preview_script }}', '');
```
Comment (aschempp) [inline core-bundle/src/Resources/views/Frontend/preview_toolbar_base_js.html.twig]:
```suggestion
                const href = window.location.href.replace('{{ preview_script|escape('js') }}', '');
```

**Review by aschempp — CHANGES_REQUESTED** [review]
Looks very good! Just two minor nitpicks 🙃

File: core-bundle/src/Resources/contao/languages/de/default.xlf [hunk 6]
```diff
@@ -2899,4 +2899,4 @@
       </trans-unit>
     </body>
   </file>
-</xliff>
\ No newline at end of file
+</xliff>
```
Comment (Toflar) [inline core-bundle/src/Resources/contao/languages/de/default.xlf]: Can you undo these changes, please? :)

**Review by ausi — DISMISSED** [review]

**Comment by simonreitinger** [comment]
Any updates on this? Have I missed something?

**Review by aschempp — DISMISSED** [review]
The only change required is what @Toflar mentioned. Please be aware that PRs are not always merged instantly if they are ready 😉

**Comment by aschempp** [comment]
Looking at the files changed in the pull request, you can see there are now TWO language files with only line ending changes. I guess you changed the wrong file (in calendar-bundle instead of core-bundle)? 🙈

**Review by m-vo — DISMISSED** [review]

**Review by aschempp — DISMISSED** [review]

**Review by Toflar — DISMISSED** [review]

**Review by ausi — DISMISSED** [review]

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @simonreitinger.
